cranelift/codegen/src/isa/s390x/inst.isle

```diff
@@ -0,0 +1,2833 @@
+
+
+;; Helpers for stack-slot addresses ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(decl abi_stackslot_addr (WritableReg StackSlot Offset32) MInst)
+(extern constructor abi_stackslot_addr abi_stackslot_addr)
+
+(decl stack_addr_impl (Type StackSlot Offset32) Reg)
+(rule (stack_addr_impl ty stack_slot offset)
+      (let ((dst WritableReg (temp_writable_reg ty))
+            (_ Unit (emit (abi_stackslot_addr dst stack_slot offset))))
+        (writable_reg_to_reg dst)))
+
+
+;; Helpers for extracting extensions ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; A value that is the result of a sign-extend from a 32-bit value.
+(decl sext32_value (Value) Value)
+(extractor (sext32_value x) (def_inst (sextend (and x (value_type $I32)))))
+
+;; A value that is the result of a zero-extend from a 32-bit value.
+(decl zext32_value (Value) Value)
+(extractor (zext32_value x) (def_inst (uextend (and x (value_type $I32)))))
+
+
+;; Helpers for sinkable loads ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; Extract a sinkable instruction from a value operand.
+(decl sinkable_inst (Inst) Value)
+(extern extractor sinkable_inst sinkable_inst)
+
+;; Sink a sinkable instruction.
+;;
+;; This is a side-effectful operation that notifies the context that the
+;; sinkable instruction been sunk into another instruction, and no longer
+;; needs to be lowered.
+(decl sink_inst (Inst) Unit)
+(extern constructor sink_inst sink_inst)
+
+
+;; Sinkable big-endian load instruction.
+(decl sinkable_load (Inst) Value)
+(extractor (sinkable_load inst)
+           (sinkable_inst (and inst (load (bigendian) _addr _offset))))
+
+;; Sinkable big-endian load instruction (32/64-bit types only).
+(decl sinkable_load_32_64 (Inst) Value)
+(extractor (sinkable_load_32_64 inst)
+           (and (value_type (ty_32_or_64 _)) (sinkable_load inst)))
+
+;; Sinkable big-endian load instruction (16-bit types only).
+(decl sinkable_load_16 (Inst) Value)
+(extractor (sinkable_load_16 inst)
+           (and (value_type $I16) (sinkable_load inst)))
+
+;; Sinkable big-endian sload16 instruction.
+(decl sinkable_sload16 (Inst) Value)
+(extractor (sinkable_sload16 inst)
+           (sinkable_inst (and inst (sload16 (bigendian) _addr _offset))))
+
+;; Sinkable big-endian sload32 instruction.
+(decl sinkable_sload32 (Inst) Value)
+(extractor (sinkable_sload32 inst)
+           (sinkable_inst (and inst (sload32 (bigendian) _addr _offset))))
+
+;; Sinkable big-endian uload16 instruction.
+(decl sinkable_uload16 (Inst) Value)
+(extractor (sinkable_uload16 inst)
+           (sinkable_inst (and inst (uload16 (bigendian) _addr _offset))))
+
+;; Sinkable big-endian uload32 instruction.
+(decl sinkable_uload32 (Inst) Value)
+(extractor (sinkable_uload32 inst)
+           (sinkable_inst (and inst (uload32 (bigendian) _addr _offset))))
+
+;; Sink a load instruction, returning its address as `MemArg`.
+(decl sink_load (Inst) MemArg)
```

Comment:
Nitpick: it would be nice if the docs for this term and the other similar ones below it mentioned that this is a side-effectful operation.

Reply:
Of course, done.